src/uucore/src/lib/mods/backup_control.rs

```diff
@@ -37,22 +33,164 @@ pub fn determine_backup_suffix(supplied_suffix: Option<&str>) -> String {
+/// backup modes.
+///
+///
+/// ```
+/// let OPT_BACKUP: &str = "backup";
+/// let OPT_BACKUP_NO_ARG: &str = "b";
+/// let matches = App::new("myprog")
+///     .arg(Arg::with_name(OPT_BACKUP_NO_ARG)
+///         .short(OPT_BACKUP_NO_ARG)
+///     .arg(Arg::with_name(OPT_BACKUP)
+///         .long(OPT_BACKUP)
+///         .takes_value(true)
+///         .require_equals(true)
+///         .min_values(0))
+///     .get_matches_from(vec![
+///         "myprog", "-b", "--backup=n"
+///     ]);
+///
+/// let backup_mode = backup_control::determine_backup_mode(
+///     matches.is_present(OPT_BACKUP_NO_ARG), matches.is_present(OPT_BACKUP),
+///     matches.value_of(OPT_BACKUP)
+/// );
+/// let backup_mode = match backup_mode {
+///     Err(err) => {
+///         show_usage_error!("{}", err);
+///         return 1;
+///     },
+///     Ok(mode) => mode,
+/// };
+/// ```
+pub fn determine_backup_mode(
+    sopt_present: bool,
+    lopt_present: bool,
+    lopt_value: Option<&str>,
+) -> Result<BackupMode, String> {
+    if lopt_present {
+        // Use method to determine the type of backups to make. When this option
+        // is used but method is not specified, then the value of the
+        // VERSION_CONTROL environment variable is used. And if VERSION_CONTROL
+        // is not set, the default backup type is ‘existing’.
+        if let Some(method) = lopt_value {
+            // Second argument is for the error string that is returned.
+            match_method(method, "backup type")
+        } else if let Ok(method) = env::var("VERSION_CONTROL") {
+            // Second argument is for the error string that is returned.
+            match_method(&method, "$VERSION_CONTROL")
+        } else {
+            Ok(BackupMode::ExistingBackup)
+        }
+    } else if sopt_present {
+        // the short form of this option, -b does not accept any argument.
+        // Using -b is equivalent to using --backup=existing.
+        Ok(BackupMode::ExistingBackup)
+    } else {
+        // No option was present at all
+        Ok(BackupMode::NoBackup)
+    }
+}
+
+/// Match a backup option string to a `BackupMode`.
+///
+/// The GNU manual specifies that abbreviations to options are valid as long as
+/// they aren't ambiguous. This function matches the given `method` argument
+/// against all valid backup options (via `starts_with`), and returns a valid
+/// [`BackupMode`] if exactly one backup option matches the `method` given.
+///
+/// `origin` is required in order to format the generated error message
+/// properly, when an error occurs.
+///
+///
+/// # Errors
+///
+/// If `method` is ambiguous (i.e. may resolve to multiple backup modes) or
+/// invalid, an error is returned. The error contains the formatted error string
+/// which may then be passed to the [`show_usage_error`] macro.
+fn match_method(method: &str, origin: &str) -> Result<BackupMode, String> {
+    let x = vec![
```

Comment:
This doesn't need to be a `Vec` but could be a plain array. (i.e. you can remove `vec!` and everything should still work.)

Reply:
Using the definitions from `BACKUP_CONTROL_VALUES` instead now, which already exists in this module.